api4/graphql.go

```diff
@@ -0,0 +1,156 @@
+	"github.com/mattermost/mattermost-server/v6/shared/mlog"
+)
+
+type graphQLInput struct {
+	Query         string                 `json:"query"`
+	OperationName string                 `json:"operationName"`
+	Variables     map[string]interface{} `json:"variables"`
+}
+
+//go:embed schema.graphqls
+var schemaRaw string
+
+func (api *API) InitGraphQL() error {
+	// Guard with a feature flag.
+	if !api.srv.Config().FeatureFlags.GraphQL {
+		return nil
+	}
+
+	var err error
+	opts := []graphql.SchemaOpt{
+		graphql.UseFieldResolvers(),
+		graphql.Logger(mlog.NewGraphQLLogger(api.srv.Log)),
+		graphql.MaxParallelism(5),
+	}
+
+	if isProd() {
+		opts = append(opts,
+			// MaxDepth cannot be moved as a general param
+			// because otherwise introspection also doesn't work
+			// with just a depth of 4.
+			graphql.MaxDepth(4),
+			graphql.DisableIntrospection(),
+		)
+	}
+
+	api.schema, err = graphql.ParseSchema(schemaRaw, &resolver{}, opts...)
+	if err != nil {
+		return err
+	}
+
+	api.BaseRoutes.APIRoot5.Handle("/graphql", api.APIHandlerTrustRequester(graphiQL)).Methods("GET")
+	api.BaseRoutes.APIRoot5.Handle("/graphql", api.APISessionRequired(api.graphQL)).Methods("POST")
+	return nil
+}
+
+// Unique type to hold our context.
+type ctxKey struct{}
+
+func (api *API) graphQL(c *Context, w http.ResponseWriter, r *http.Request) {
+	var response *graphql.Response
+	defer func() {
+		if response != nil {
+			if err := json.NewEncoder(w).Encode(response); err != nil {
+				mlog.Warn("Error while writing response", mlog.Err(err))
+			}
+		}
+	}()
+
+	// Limit bodies to 100KiB.
+	// We need to enforce a lower limit than the file upload size,
+	// to prevent the library doing unnecessary parsing.
+	r.Body = http.MaxBytesReader(w, r.Body, 102400)
+
+	var params graphQLInput
+	if err := json.NewDecoder(r.Body).Decode(&params); err != nil {
+		err2 := gqlerrors.Errorf("invalid request body: %v", err)
+		response = &graphql.Response{Errors: []*gqlerrors.QueryError{err2}}
+		return
+	}
+
+	if isProd() && params.OperationName == "" {
+		err2 := gqlerrors.Errorf("operation name not passed")
+		response = &graphql.Response{Errors: []*gqlerrors.QueryError{err2}}
+		return
+	}
+
+	// Populate the context with required info.
+	reqCtx := r.Context()
+	reqCtx = context.WithValue(reqCtx, ctxKey{}, c)
+
+	response = api.schema.Exec(reqCtx,
+		params.Query,
+		params.OperationName,
+		params.Variables)
+
+	if len(response.Errors) > 0 {
+		mlog.Error("Error executing request", mlog.String("operation", params.OperationName),
+			mlog.Array("errors", response.Errors))
+	}
+}
+
+func graphiQL(c *Context, w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", "text/html")
+	w.Write(graphiqlPage)
+}
+
+var graphiqlPage = []byte(`
+<!DOCTYPE html>
+<html>
+	<head>
+		<title>GraphiQL editor | Mattermost</title>
+		<link href="https://cdnjs.cloudflare.com/ajax/libs/graphiql/0.11.11/graphiql.min.css" rel="stylesheet" />
```

Comment:
Hmm, why we are not hosting these with the app server?

Reply:
Not a bad idea, it's just that this was the simplest way to implement it. Created: https://mattermost.atlassian.net/browse/MM-41614